File: apps/user-desktop/src/lib/adapters/window.ts
/*
 * SPDX-License-Identifier: AGPL-3.0-only
 * Copyright (C) 2026 baibai and Botting contributors
 *
 * Botting is free software: you can redistribute it and/or modify it under
 * the GNU Affero General Public License version 3, as published by the
 * Free Software Foundation. This program comes WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the LICENSE file for the complete terms.
 * Copyleft: covered modifications must retain these license obligations.
 * https://www.gnu.org/licenses/agpl-3.0.html
 */

/** 協調原生視窗、WebView 縮放及檔案拖放；版面切換會依序完成。 */
import { invoke } from "@tauri-apps/api/core";
import { LogicalSize } from "@tauri-apps/api/dpi";
import { getCurrentWebview } from "@tauri-apps/api/webview";
import { currentMonitor, getCurrentWindow, type DragDropEvent } from "@tauri-apps/api/window";
import { fitWindowToWorkArea } from "../../../../shared-ui/window-fit";
import {
  canvasScaleForCurrentViewport,
  waitForViewportLayout,
} from "../../../../shared-ui/viewport-fit";

const isTauri = typeof window !== "undefined" && "__TAURI_INTERNALS__" in window;

export type UserWindowMode = "entry" | "app";

const entryWindowSize = { width: 514, height: 350 } as const;
const appWindowSize = { width: 1180, height: 760 } as const;
const windowEdgeMargin = 24;

interface PhysicalWindowBounds {
  x: number;
  y: number;
  width: number;
  height: number;
}

interface WindowTarget {
  bounds: PhysicalWindowBounds | null;
  size: LogicalSize;
}

function setEntrySurfaceScale(): void {
  const scaleX = entryWindowSize.width / appWindowSize.width;
  const scaleY = entryWindowSize.height / appWindowSize.height;
  const rootStyle = document.documentElement.style;
  rootStyle.setProperty("--app-entry-start-scale-x", String(scaleX));
  rootStyle.setProperty("--app-entry-start-scale-y", String(scaleY));
  rootStyle.setProperty("--app-entry-start-radius-x", `${18 / scaleX}px`);
  rootStyle.setProperty("--app-entry-start-radius-y", `${18 / scaleY}px`);
}

function fitCanvasToViewport(): void {
  const isEntry = document.documentElement.dataset.userWindow === "entry";
  const designSize = isEntry ? entryWindowSize : appWindowSize;
  const property = isEntry ? "--entry-canvas-scale" : "--user-canvas-scale";
  document.documentElement.style.setProperty(
    property,
    String(canvasScaleForCurrentViewport(designSize)),
  );
}

async function resetWebviewZoom(): Promise<void> {
  try {
    await getCurrentWebview().setZoom(1);
  } catch (error) {
    console.warn(
      "Unable to reset the user WebView zoom; fitting the canvas to the viewport.",
      error,
    );
  }
}

/**
 * 依可見範圍同步畫布縮放，並訂閱尺寸變化。
 * @return 解除 resize 監聽的清理函式。
 */
export function watchUserCanvas(): () => void {
  const update = () => fitCanvasToViewport();
  update();
  window.addEventListener("resize", update);
  window.visualViewport?.addEventListener("resize", update);
  return () => {
    window.removeEventListener("resize", update);
    window.visualViewport?.removeEventListener("resize", update);
  };
}

async function fitWindowTarget(
  designSize: typeof entryWindowSize | typeof appWindowSize,
): Promise<WindowTarget> {
  const monitor = await currentMonitor();
  const workArea = monitor?.workArea.size.toLogical(monitor.scaleFactor);
  const fitted = fitWindowToWorkArea(designSize, workArea, windowEdgeMargin);
  const size = new LogicalSize(fitted.size.width, fitted.size.height);
  const physicalSize = monitor ? size.toPhysical(monitor.scaleFactor) : null;
  const width = physicalSize ? Math.round(physicalSize.width) : 0;
  const height = physicalSize ? Math.round(physicalSize.height) : 0;
  return {
    bounds: monitor
      ? {
          x: monitor.workArea.position.x + Math.round((monitor.workArea.size.width - width) / 2),
          y: monitor.workArea.position.y + Math.round((monitor.workArea.size.height - height) / 2),
          width,
          height,
        }
      : null,
    size,
  };
}

// 將原生版面操作串接，避免相鄰切換互相覆蓋尺寸和可見狀態。
let windowModeSequence = Promise.resolve();
/**
 * 依序調整視窗、WebView 與畫布縮放，避免相鄰切換覆寫。
 * @param mode entry 或 app 視窗模式。
 * @return 版面切換完成的 Promise。
 */
export async function setUserWindowMode(mode: UserWindowMode): Promise<void> {
  if (!isTauri) {
    document.documentElement.dataset.userWindow = mode === "app" ? "main" : "entry";
    fitCanvasToViewport();
    return;
  }

  windowModeSequence = windowModeSequence
    .catch(() => undefined)
    .then(async () => {
      const target = await fitWindowTarget(mode === "app" ? appWindowSize : entryWindowSize);
      const appWindow = getCurrentWindow();
      await resetWebviewZoom();
      if (mode === "app") {
        setEntrySurfaceScale();
      }
      if (mode === "entry") {
        await appWindow.setSize(target.size);
        await appWindow.center();
      } else if (target.bounds) {
        await invoke("set_main_window_layout", {
          x: target.bounds.x,
          y: target.bounds.y,
          width: target.bounds.width,
          height: target.bounds.height,
        });
      } else {
        await appWindow.setSize(target.size);
        await appWindow.center();
      }
      await waitForViewportLayout();
      document.documentElement.dataset.userWindow = mode === "app" ? "main" : "entry";
      fitCanvasToViewport();
      await appWindow.unminimize();
      await appWindow.show();
    });
  await windowModeSequence;
}

/**
 * 最小化目前 Tauri 視窗。
 * @return 完成的 Promise；瀏覽器環境直接返回。
 */
export async function minimizeWindow(): Promise<void> {
  if (isTauri) await getCurrentWindow().minimize();
}

/**
 * 提出目前視窗的關閉要求，由後端負責停止工作階段。
 * @return 提出要求的 Promise。
 */
export async function closeWindow(): Promise<void> {
  if (isTauri) await getCurrentWindow().close();
}

/**
 * 訂閱 Tauri 原生拖放事件。
 * @param handler 拖放階段、位置及路徑的處理回呼。
 * @return 解除監聽函式的 Promise。
 */
export async function onFileDrop(
  handler: (event: DragDropEvent) => void | Promise<void>,
): Promise<() => void> {
  if (!isTauri) return () => undefined;
  return getCurrentWindow().onDragDropEvent(({ payload }) => handler(payload));
}

export type FileDropEvent = DragDropEvent;

File: apps/user-desktop/src-tauri/src/commands/window.rs
/*
 * SPDX-License-Identifier: AGPL-3.0-only
 * Copyright (C) 2026 baibai and Botting contributors
 *
 * Botting is free software: you can redistribute it and/or modify it under
 * the GNU Affero General Public License version 3, as published by the
 * Free Software Foundation. This program comes WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the LICENSE file for the complete terms.
 * Copyleft: covered modifications must retain these license obligations.
 * https://www.gnu.org/licenses/agpl-3.0.html
 */

//! 同步調整主視窗與 WebView 的邊界，避免展開動畫期間畫布尺寸落後。

use crate::app_state::CommandResult;
const MAIN_WINDOW_LABEL: &str = "main";

#[derive(Clone, Copy)]
struct WindowBounds {
    x: i32,
    y: i32,
    width: i32,
    height: i32,
}

#[tauri::command]
/// 驗證主視窗及正尺寸後同步調整原生視窗與 WebView。
/// @param window 提出要求的主視窗。
/// @param x 實體像素左座標。
/// @param y 實體像素上座標。
/// @param width 實體像素寬度。
/// @param height 實體像素高度。
/// @return 視窗與 WebView 更新結果。
pub(crate) async fn set_main_window_layout(
    window: tauri::WebviewWindow,
    x: i32,
    y: i32,
    width: i32,
    height: i32,
) -> CommandResult<()> {
    if window.label() != MAIN_WINDOW_LABEL {
        return Err("window bounds can only be changed for the main window".into());
    }
    if width <= 0 || height <= 0 {
        return Err("window dimensions must be positive".into());
    }
    let target = WindowBounds {
        x,
        y,
        width,
        height,
    };

    set_platform_window_layout(window, target).await
}

#[cfg(windows)]
async fn set_platform_window_layout(
    window: tauri::WebviewWindow,
    target: WindowBounds,
) -> CommandResult<()> {
    let hwnd = window.hwnd().map_err(|error| error.to_string())?.0 as usize;
    let (sender, receiver) = tokio::sync::oneshot::channel();

    window
        .with_webview(move |webview| {
            let result = set_windows_webview_layout(webview, hwnd, target);
            let _ = sender.send(result);
        })
        .map_err(|error| error.to_string())?;

    receiver
        .await
        .map_err(|_| "window closed before its bounds could be changed".to_string())?
}

#[cfg(windows)]
fn set_windows_webview_layout(
    webview: tauri::webview::PlatformWebview,
    hwnd: usize,
    target: WindowBounds,
) -> CommandResult<()> {
    use std::{io, ptr};
    use windows_sys::Win32::Graphics::Gdi::{
        RedrawWindow, RDW_ALLCHILDREN, RDW_ERASE, RDW_INVALIDATE, RDW_UPDATENOW,
    };
    use windows_sys::Win32::UI::WindowsAndMessaging::{
        GetWindowRect, SetWindowPos, SWP_NOACTIVATE, SWP_NOCOPYBITS, SWP_NOOWNERZORDER,
        SWP_NOZORDER,
    };

    let hwnd = hwnd as windows_sys::Win32::Foundation::HWND;
    let mut current_bounds = Default::default();
    // SAFETY: hwnd 來自仍存活的 Tauri 主視窗，current_bounds 指向可寫入的結構。
    let has_current_bounds = unsafe { GetWindowRect(hwnd, &mut current_bounds) } != 0;
    let bounds_changed = !has_current_bounds
        || current_bounds.left != target.x
        || current_bounds.top != target.y
        || current_bounds.right - current_bounds.left != target.width
        || current_bounds.bottom - current_bounds.top != target.height;

    if bounds_changed {
        let controller = webview.controller();
        let mut webview_bounds = Default::default();
        // SAFETY: controller 由 WebView2 持有，邊界先初始化再使用，視窗尺寸已在指令入口驗證。
        let succeeded = unsafe {
            controller
                .Bounds(&mut webview_bounds)
                .map_err(|error| error.to_string())?;
            webview_bounds.left = 0;
            webview_bounds.top = 0;
            webview_bounds.right = target.width;
            webview_bounds.bottom = target.height;
            controller
                .SetBounds(webview_bounds)
                .map_err(|error| error.to_string())?;

            SetWindowPos(
                hwnd,
                ptr::null_mut(),
                target.x,
                target.y,
                target.width,
                target.height,
                SWP_NOCOPYBITS | SWP_NOACTIVATE | SWP_NOOWNERZORDER | SWP_NOZORDER,
            )
        };
        if succeeded == 0 {
            return Err(io::Error::last_os_error().to_string());
        }
        // SAFETY: controller 屬於剛完成邊界更新且仍存活的同一個 WebView。
        unsafe {
            controller
                .NotifyParentWindowPositionChanged()
                .map_err(|error| error.to_string())?;
        }
    }

    // SAFETY: RedrawWindow 只要求仍存活的主視窗與子視窗重新繪製。
    unsafe {
        RedrawWindow(
            hwnd,
            ptr::null(),
            ptr::null_mut(),
            RDW_INVALIDATE | RDW_ERASE | RDW_UPDATENOW | RDW_ALLCHILDREN,
        );
    }
    Ok(())
}

#[cfg(not(windows))]
async fn set_platform_window_layout(
    window: tauri::WebviewWindow,
    target: WindowBounds,
) -> CommandResult<()> {
    window
        .set_position(tauri::PhysicalPosition::new(target.x, target.y))
        .map_err(|error| error.to_string())?;
    window
        .set_size(tauri::PhysicalSize::new(
            u32::try_from(target.width).map_err(|_| "window width must be positive")?,
            u32::try_from(target.height).map_err(|_| "window height must be positive")?,
        ))
        .map_err(|error| error.to_string())
}
